### PR TITLE
Update download dependencies

### LINKED ISSUE #1518: Can't compile: Unable to detect OpenSSL version
I'm getting this trying to compile 1.14.0:

```
error: failed to run custom build command for `openssl v0.9.24`
process didn't exit successfully: `/build/rustup/src/rustup.rs-1.14.0/target/release/build/openssl-4e050bcc8f5aae18/build-script-build` (exit code: 101)
--- stderr
thread 'main' panicked at 'Unable to detect OpenSSL version', /build/.cargo/registry/src/github.com-1ecc6299db9ec823/openssl-0.9.24/build.rs:16:14
note: Run with `RUST_BACKTRACE=1` for a backtrace.

warning: build failed, waiting for other jobs to finish...
error: build failed
```

My rust is 1.29.0 and my openssl is 1.1.1. This is on Arch Linux.

### PR DESCRIPTION
**Pull request by pickfire**
- reqwest 0.9, fix #1518
- hyper 0.12

### CODE REVIEW

File: src/download/src/lib.rs [hunk 1]
```diff
@@ -277,8 +277,10 @@ pub mod reqwest_be {
         let buffer_size = 0x10000;
         let mut buffer = vec![0u8; buffer_size];
 
-        if let Some(len) = res.headers().get::<header::ContentLength>() {
-            callback(Event::DownloadContentLengthReceived(len.0 + resume_from))?;
+        if let Some(len) = res.headers().get(header::CONTENT_LENGTH) {
+            // TODO possible issues during unwrap?
+            let len = len.to_str().unwrap().parse::<u64>().unwrap() + resume_from;
```
Comment (meithecatte) [inline src/download/src/lib.rs:282]: What is the purpose of converting the header to a string and back?

Comment (pickfire) [reply src/download/src/lib.rs:282]: `len` is `&reqwest::header::HeaderValue` (not `u64`) there IIRC so I need to convert it to `str` and then parse it as `u64`.

Even though `HeaderValue` have `HttpTryFrom` but `TryFrom` is nigthly so not usable.

I removing the extra `to_str` but `HeaderValue` does not implement `parse`.

Maybe I might have missed something?

**Comment by ghost** [comment]
**tl;dr**: ignore this. Just hitting https://github.com/rust-lang/rust/issues/55376

Changing just [this](https://github.com/rust-lang-nursery/rustup.rs/pull/1523/files#diff-abf3409f55470717066a1d80cdfebd82L22) line in file [src/download/Cargo.toml](https://github.com/rust-lang-nursery/rustup.rs/blob/9774f514ce3630bbd6f0b77ca89a8431cc4fb4f4/src/download/Cargo.toml)  (with a checked out rustup.rs (latest) commit being at 3abd09778e7185faab7b5c4d0c7fec81b0674388)

```patch
-reqwest = { version = "0.8.4", optional = true }
+reqwest = { version = "0.9", optional = true }
```
causes an ICE

```
$ ./go
!! LD_LIBRARY_PATH=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib:
!! Executing '/home/xftroxgpx/.cargo/bin/cargo' in pwd='/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustup.rs/rustup.rs' with args: 'install --force'
warning: Using `cargo install` to install the binaries for the package in current working directory is deprecated, use `cargo install --path .` instead. Use `cargo build` if you want to simply build the package.
  Installing rustup v1.14.0 (/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustup.rs/rustup.rs)
   Compiling hyper v0.12.13                                                     
thread 'main' panicked at 'librustc/hir/map/hir_id_validator.rs:31:             
HirIdValidator: The recorded owner of path segment super (id=36819) is ::server[0]::conn[0]::{{?}}[34] instead of ::server[0]::conn[0]::{{?}}[34]::{{?}}[0]
HirIdValidator: Same HirId ::server[0]::conn[0]::{{?}}[34]::{{?}}[0]/2 assigned for nodes path segment super (id=36819) and path segment spawn_all (id=88814)', librustc/util/bug.rs:47:26
note: Run with `RUST_BACKTRACE=1` for a backtrace.                              
                                                                                
error: internal compiler error: unexpected panic                                
                                                                                
note: the compiler unexpectedly panicked. this is a bug.                        
                                                                                
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
                                                                                
note: rustc 1.31.0-nightly (cae6efc37 2018-10-27) running on x86_64-unknown-linux-gnu
                                                                                
note: compiler flags: -C opt-level=3 --crate-type lib                           
                                                                                
note: some of the compiler flags provided by cargo are hidden                   
                                                                                
error: failed to compile `rustup v1.14.0 (/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustup.rs/rustup.rs)`, intermediate artifacts can be found at `/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustup.rs/rustup.rs/target`

Caused by:
  Could not compile `hyper`.

To learn more, run the command again with --verbose.
```

```
$ rustv
!! LD_LIBRARY_PATH=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib:
!! Executing '/home/xftroxgpx/.cargo/bin/rustc' in pwd='/home/xftroxgpx' with args: '-vV'
rustc 1.31.0-nightly (cae6efc37 2018-10-27)
binary: rustc
commit-hash: cae6efc37d70ab7d353e6ab9ce229d59a65ed643
commit-date: 2018-10-27
host: x86_64-unknown-linux-gnu
release: 1.31.0-nightly
LLVM version: 8.0
!! LD_LIBRARY_PATH=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib:
!! Executing '/home/xftroxgpx/.cargo/bin/cargo' in pwd='/home/xftroxgpx' with args: '-vV'
cargo 1.31.0-nightly (2d0863f65 2018-10-20)
release: 1.31.0
commit-hash: 2d0863f657e6f45159fc7412267eee3e659185e5
commit-date: 2018-10-20
```


Ah, it's already reported: https://github.com/rust-lang/rust/issues/55376

**Comment by ghost** [comment]
I've just tested this PR and it fixes #1518 for me.
Cheers!


some unrelated warnings though:
```
   Compiling reqwest v0.9.2                                                     
   Compiling download v0.4.0 (/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustup.rs/rustup.rs/src/download)
   Compiling rustup-utils v1.14.0 (/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustup.rs/rustup.rs/src/rustup-utils)
warning: use of deprecated item 'std::env::home_dir': This function's behavior is unexpected and probably not what you want. Consider using the home_dir function from https://crates.io/crates/dirs instead.
   --> src/rustup-utils/src/utils.rs:519:5                                      
    |                                                                           
519 |     ::std::env::home_dir()                                                
    |     ^^^^^^^^^^^^^^^^^^^^                                                  
    |                                                                           
    = note: #[warn(deprecated)] on by default                                   
                                                                                
   Compiling rustup-dist v1.14.0 (/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustup.rs/rustup.rs/src/rustup-dist)
   Compiling rustup v1.14.0 (/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustup.rs/rustup.rs)
warning: use of deprecated item 'std::env::home_dir': This function's behavior is unexpected and probably not what you want. Consider using the home_dir function from https://crates.io/crates/dirs instead.
   --> src/rustup-cli/self_update.rs:388:31                                     
    |                                                                           
388 |     let multirust_meta_path = env::home_dir().map(|d| d.join(".multirust"));
    |                               ^^^^^^^^^^^^^                               
    |                                                                           
    = note: #[warn(deprecated)] on by default                                   
                                                                                
warning: use of deprecated item 'std::env::home_dir': This function's behavior is unexpected and probably not what you want. Consider using the home_dir function from https://crates.io/crates/dirs instead.
   --> src/rustup-cli/self_update.rs:390:26                                     
    |                                                                           
390 |     let rustup_sh_path = env::home_dir().map(|d| d.join(".rustup"));      
    |                          ^^^^^^^^^^^^^                                    
                                                                                
    Finished release [optimized] target(s) in 4m 11s     
```

**Comment by pickfire** [comment]
@xftroxgpx Yup, I am already planning to do those after this commit is merge. Still waiting for the pull request to be merged. And maybe reformat the project and reorganize some tests as well.

**Comment by svenstaro** [comment]
Can we get this merged? Without it, rustup doesn't build on modern systems.

**Comment by nrc** [comment]
Thanks for the PR, sorry for the delay in merging.

Please note though that Rustup reviewers do not have much available bandwidth right not, so if anyone is keen on implementing more, please check that someone will be able to review it. (More info at https://internals.rust-lang.org/t/cargo-and-rustup-development)

**Comment by svenstaro** [comment]
@nrc I understand that you guys don't have a lot of time right now but could you make a release with this fix? It's quite important because more and more distros will start offering OpenSSL 1.1 by default and then it just won't build.

**Comment by nrc** [comment]
> I understand that you guys don't have a lot of time right now but could you make a release with this fix?

I'm planning to make a Rustup release in a couple of weeks time. I hope that is soon enough (because we don't have release trains, rustup releases are a fair bit of work and are somewhat risky).
